dashboard/app.py

```diff
@@ -68,22 +68,42 @@ def main():
 
 @app.route('/leaderboard')
 def leaderboard():
-    page_content = cache.get('leaderboard')
+    page = int(request.args.get('page', 1))
+    if page < 1:
+        abort(400)
+
+    page_content = cache.get('leaderboard-page-{}'.format(page))
     if page_content is None:
+        offset = (page - 1) * LEADERBOARD_ROWS_PER_PAGE
         date_from, date_to = get_week_range(datetime.utcnow().date())
-        leader_board_rows = get_leaderboard_rows(date_from, date_to, 10)
-        enrich_leaderboard_rows(leader_board_rows, date_from, date_to)
+        leaderboard_rows = get_leaderboard_rows(
+            date_from, date_to, offset=offset, limit=LEADERBOARD_ROWS_PER_PAGE
+        )
+        enrich_leaderboard_rows(leaderboard_rows, date_from, date_to)
+
+        previous_page = None
+        if page >= 2:
+            previous_page = page - 1
+
+        next_page = None
+        if len(leaderboard_rows) == LEADERBOARD_ROWS_PER_PAGE:
+            next_page = page + 1
+
+        index_start = (page - 1) * LEADERBOARD_ROWS_PER_PAGE
         page_content = render_template(
             'leaderboard.html',
-            date_from=date_from.strftime("%Y-%m-%d"),
-            date_to=date_to.strftime("%Y-%m-%d"),
-            leaderboard_rows=leader_board_rows,
-        )
-        cache.set(
-            'leaderboard',
-            page_content,
-            timeout=1 * 60
+            date_from=date_from.strftime('%b %d, %Y'),
+            date_to=date_to.strftime('%b %d, %Y'),
+            leaderboard_rows=leaderboard_rows,
+            previous_page=previous_page,
+            next_page=next_page,
+            index_start=index_start
         )
+    cache.set(
+        'leaderboard-page-{}'.format(page),
```

Comment:
Caching every page separately will cause interesting bugs, because cash timeouts in 5 minutes.
Caches of different pages will invalidate at different times, so node rows will be duplicated across multiple pages.

Example: you'll see yourself in 10th place (because you were 9th 4 minutes ago) and also in 11th place in 2nd page (because now you're actually 10th).

Solution for this might be to calculate leaderboard for all rows at once, but keep only models in cache: `cache.set('leaderboard-rows', leaderboard_rows)`

But this causes another issue - if there are 2000 nodes, calculating rows for all of them sucks :/

So it's probably good as it is - these are just some thoughts around future problems :D